### PR TITLE
Update breadcrumbs

### PR DESCRIPTION
**Pull request by Yash621**
updated breadcrumbs navigation structure

demo:

assets:

image

plan:

image

### CODE REVIEW

**Review by peacog — CHANGES_REQUESTED** [review]
Hi @Yash621 this is a good start but there is more to do. 

1. Remove the **Back to site** element. It's confusing and not at all useful. Replace it with Home.
2. Change the last element of the breadcrumb, the one that says Edit Asset and Edit Plan, to the name of the asset or plan. With this MR you've added an extra element that's a link to the page you're on. We don't want that because it's confusing. The breadcrumb for a CFR asset should look like this ``Home / Records / Locations / CFRs / Abera``, and Abera should not be clickable.
3. With a breadcrumb trail it should be possible to click on any level of the breadcrumb and see the trail up to that point. For example if you're on Abera the breadcrumb is: ``Home / Records / Locations / CFRs / Abera`` If you click on CFRs that page should have the following breadcrumb: ``Home / Records / Locations / CFRs`` and if you click on Locations you should have ``Home / Records / Locations``
4. The range and sector breadcrumbs should be ``Home / Records / Locations / Sectors`` and ``Home / Records / Locations / Ranges``. Currently they have 'Land' instead of 'Sector' and 'Range'
5. The breadcrumb should work on asset tabs, i.e. the GFW, Management, Logs, Assets tabs on CFRs and the tabs on the other asset types. For example the GFW tab on Abera should have the breadcrumb ``Home / Records / Locations / CFRs / Abera / GFW`` and the Abera element should be a link.
6. The above comments apply to Plans too.

**Comment by Yash621** [comment]
> Hi @Yash621 this is a good start but there is more to do.
> 
> 1. Remove the **Back to site** element. It's confusing and not at all useful. Replace it with Home.
> 2. Change the last element of the breadcrumb, the one that says Edit Asset and Edit Plan, to the name of the asset or plan. With this MR you've added an extra element that's a link to the page you're on. We don't want that because it's confusing. The breadcrumb for a CFR asset should look like this `Home / Records / Locations / CFRs / Abera`, and Abera should not be clickable.
> 3. With a breadcrumb trail it should be possible to click on any level of the breadcrumb and see the trail up to that point. For example if you're on Abera the breadcrumb is: `Home / Records / Locations / CFRs / Abera` If you click on CFRs that page should have the following breadcrumb: `Home / Records / Locations / CFRs` and if you click on Locations you should have `Home / Records / Locations`
> 4. The range and sector breadcrumbs should be `Home / Records / Locations / Sectors` and `Home / Records / Locations / Ranges`. Currently they have 'Land' instead of 'Sector' and 'Range'
> 5. The breadcrumb should work on asset tabs, i.e. the GFW, Management, Logs, Assets tabs on CFRs and the tabs on the other asset types. For example the GFW tab on Abera should have the breadcrumb `Home / Records / Locations / CFRs / Abera / GFW` and the Abera element should be a link.
> 6. The above comments apply to Plans too.

yup @peacog thanks for tagging these here, i am aware and currently working on fixing the listed issues, will let you know once it's fixed!

**Review by peacog — CHANGES_REQUESTED** [review]
Hi @Yash621 I've left you some comments in the code. Also some of the breadcrumbs are wrong:

1. https://forests.ddev.site/assets the last element is 'Pages'
2. On the asset page the asset type should be in plural i.e. CFRs, Sectors, Ranges. That0s because the parent level shows the plural version. They naming should be consistent
3. On assets the breadcrumb is wrong on the Logs and Assets tabs
4. On the plan page the Permissions breadcrumb is Edit-permissions. Change that to Permissions. you can change it in farm_nfa_entity_type_alter and farm_nfa.routing.yml

File: web/modules/custom/farm_nfa/src/BreadcrumbBuilder.php [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    $breadcrumb = new Breadcrumb();
+    $links = [];
+    $asset = $route_match->getParameter('asset');
+    $plan = $route_match->getParameter('plan');
+    $breadcrumb->addCacheContexts(['url.path']);
+    $view = $route_match->getParameter('view_id');
+    $links[] = Link::createFromRoute('Home', '<front>');
+    $current_tab = $this->getCurrentTab();
+    if ($asset instanceof EntityInterface || is_numeric($asset)) {
+      $asset = is_numeric($asset) ? \Drupal::entityTypeManager()->getStorage('asset')->load($asset) : $asset;
```
Comment (peacog) [inline web/modules/custom/farm_nfa/src/BreadcrumbBuilder.php:47]: Inject the entityTypeManager. See CommentBreadcrumbBuilder.php and comment.services.yml for an example

File: web/modules/custom/farm_nfa/src/BreadcrumbBuilder.php [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+use Drupal\Core\Link;
+use Drupal\Core\Routing\RouteMatchInterface;
+use Drupal\plan\Entity\PlanInterface;
+
+/**
+ * Breadcrumb builder class for generating breadcrumbs.
+ *
+ * Builds breadcrumbs within the context of the abv_app Drupal package.
+ *
+ * @package Drupal\abv_app
```
Comment (peacog) [inline web/modules/custom/farm_nfa/src/BreadcrumbBuilder.php:17]: remove @package and change to comment to, for example. Builds the breadcrumbs for assets and plans.

File: web/modules/custom/farm_nfa/src/BreadcrumbBuilder.php [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+    $current_tab = $this->getCurrentTab();
+    if ($asset instanceof EntityInterface || is_numeric($asset)) {
+      $asset = is_numeric($asset) ? \Drupal::entityTypeManager()->getStorage('asset')->load($asset) : $asset;
+      // If the asset parameter is a valid entity.
+      $asset_type = $asset->bundle();
+      if ($asset_type == 'land') {
+        // If the asset type is 'land', further refine it based on land type as land type can be either 'Sector' or 'Range'.
+        $asset_type = $asset->get('land_type')->value;
+      }
+      // Add a link to 'Records' for consistency in breadcrumb.
```
Comment (peacog) [inline web/modules/custom/farm_nfa/src/BreadcrumbBuilder.php:54]: change this to a @todo to remind us we need to make a Records landing page. for example
// @todo Records should link to a Records landing page, not the home page.

File: web/modules/custom/farm_nfa/src/BreadcrumbBuilder.php [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+      $links[] = Link::createFromRoute('Records', '<front>');
+      $links[] = Link::createFromRoute('Locations', 'entity.asset.collection');
+      $asset_type = $asset_type == 'cfr' ? strtoupper($asset_type) : ucfirst($asset_type);
+      $links[] = Link::createFromRoute($asset_type . 's', 'farm_nfa.assets.asset', ['asset' => strtolower($asset_type)]);
```
Comment (peacog) [inline web/modules/custom/farm_nfa/src/BreadcrumbBuilder.php:58]: Explain what we're doing here, e.g.
// The breadcrumb on the assets overview page is plural, so we pluralize
// the fragment on the individual asset page for consistency.

File: web/modules/custom/farm_nfa/src/BreadcrumbBuilder.php [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+    }
+    elseif ($view == 'farm_asset') {
+      // If the view is related to farm assets.
+      $view_type = $route_match->getParameter('arg_0') ? $route_match->getParameter('arg_0') : $route_match->getParameter('display_id');
+      $links[] = Link::createFromRoute('Records', '<front>');
+      // If there is a specific view type, add it to the breadcrumb.
+      if ($view_type && $view_type != 'page') {
+        $view_type = $view_type == 'cfr' ? strtoupper($view_type) : ucfirst($view_type);
+        $links[] = Link::createFromRoute('Locations', 'entity.asset.collection');
+        $links[] = Link::createFromRoute($view_type . 's', '<none>');
```
Comment (peacog) [inline web/modules/custom/farm_nfa/src/BreadcrumbBuilder.php:93]: As above, explain why we're adding an 's'

**Review by peacog — CHANGES_REQUESTED** [review]
It's working well now @Yash621. I requested a few more changes. 
1. We should always inject services, so we should inject the entityTypeManager. 
2. Some comment improvements
Thanks!

**Pull request closed without merging**